app/models/spree/product_set.rb

```diff
@@ -43,16 +57,18 @@ def update_product(found_model, attributes)
   def update_product_only_attributes(product, attributes)
     variant_related_attrs = [:id, :variants_attributes, :master_attributes]
     product_related_attrs = attributes.except(*variant_related_attrs)
-
     return true if product_related_attrs.blank?
 
+    original_supplier = product.supplier_id
     product.assign_attributes(product_related_attrs)
 
     product.variants.each do |variant|
       validate_presence_of_unit_value(product, variant)
     end
+    return false unless product.errors.empty? && product.save
 
-    product.save if errors.empty?
+    ExchangeVariantDeleter.new.delete(product) if original_supplier != product.supplier_id
+    true
```

Comment:
this is rubocop (use guard clause) forcing me to write confusing code.. I think this would be easier to understand:
```ruby
    if product.errors.empty? && product.save
      ExchangeVariantDeleter.new.delete(product) if original_supplier != product.supplier_id
      return true
    else
      return false
    end
```

Reply:
Maybe the deletion is in the wrong method? What about his one?

```diff
  def update_product(found_model, attributes)
+   oc_updater = ExchangeVariantDeleter.new(found_model) # caching the current supplier id
    update_product_only_attributes(found_model, attributes) &&
+     oc_updater.delete_if_supplier_changed &&
      update_product_variants(found_model, attributes) &&
      update_product_master(found_model, attributes)
  end
```

There are actually two ideas in here. One is separating updating-the-product and deleting-the-variants to simplify boolean logic and avoid branches with returns. This can also happen with a new method.

The other idea was to move the supplier storage into the service but it doesn't actually save any lines of code. So I don't think you should do that one but I wanted to share the idea.

Reply:
ah, this is interesting, thanks  for sharing!
in new commit, I moved it to update_product as suggested :+1: but keeping the logic as is because the delete method may not return truthy and we still want to run the following update methods.